main.go

```diff
@@ -616,6 +308,499 @@ func main() {
+}
+
+// Create autorizer for a tenant.
+func newAuthorizer(tCfg *tenantsConfig, t *tenant) error {
+	skip := level.Warn(log.With(tCfg.logger, "msg", "skipping invalid tenant"))
+
+	if t.OPA != nil {
+		if t.OPA.URL != "" {
+			u, err := url.Parse(t.OPA.URL)
+			if err != nil {
+				skip.Log("tenant", t.Name, "err", fmt.Sprintf("failed to parse OPA URL: %v", err))
+
+				return err
+			}
+
+			t.OPA.authorizer = opa.NewRESTAuthorizer(u, opa.LoggerOption(log.With(tCfg.logger, "tenant", t.Name)))
+		} else {
+			a, err := opa.NewInProcessAuthorizer(t.OPA.Query, t.OPA.Paths, opa.LoggerOption(log.With(tCfg.logger, "tenant", t.Name)))
+			if err != nil {
+				skip.Log("tenant", t.Name, "err", fmt.Sprintf("failed to create in-process OPA authorizer: %v", err))
+
+				return err
+			}
+			t.OPA.authorizer = a
+		}
+	}
+
+	return nil
+}
+
+// Load MTLS auth information for a tenant.
+func loadMTLSConf(tCfg *tenantsConfig, t *tenant) error {
+	var err error
+
+	skip := level.Warn(log.With(tCfg.logger, "msg", "skipping invalid tenant"))
+
+	if t.MTLS != nil {
+		if t.MTLS.CAPath != "" {
+			t.MTLS.RawCA, err = ioutil.ReadFile(t.MTLS.CAPath)
+
+			if err != nil {
+				skip.Log("tenant", t.Name, "err", fmt.Sprintf("cannot read CA certificate file for tenant %q: %v", t.Name, err))
+
+				return err
+			}
+		}
+
+		var (
+			block *pem.Block
+			rest  []byte = t.MTLS.RawCA
+			cert  *x509.Certificate
+		)
+
+		for {
+			block, rest = pem.Decode(rest)
+
+			if block == nil {
+				skip.Log("tenant", t.Name, "err", "failed to parse CA certificate PEM")
+
+				break
+			}
+
+			cert, err = x509.ParseCertificate(block.Bytes)
+
+			if err != nil {
+				skip.Log("tenant", t.Name, "err", fmt.Sprintf("failed to parse CA certificate: %v", err))
+
+				break
+			}
+
+			t.MTLS.cas = append(t.MTLS.cas, cert)
+
+			if len(rest) == 0 {
+				break
+			}
+		}
+	}
+
+	return nil
+}
+
+// Load OIDC auth information for a tenant.
+func loadOIDCConf(tCfg *tenantsConfig, t *tenant) error {
+	skip := level.Warn(log.With(tCfg.logger, "msg", "skipping invalid tenant"))
+
+	var err error
+
+	if t.OIDC != nil {
+		if t.OIDC.IssuerCAPath != "" {
+			t.OIDC.IssuerRawCA, err = ioutil.ReadFile(t.OIDC.IssuerCAPath)
+			if err != nil {
+				skip.Log("tenant", t.Name, "err", fmt.Sprintf("cannot read issuer CA certificate file for tenant : %q %v", t.Name, err))
+
+				return err
+			}
+		}
+
+		if len(t.OIDC.IssuerRawCA) != 0 {
+			block, _ := pem.Decode(t.OIDC.IssuerRawCA)
+
+			if block == nil {
+				skip.Log("tenant", t.Name, "err", "failed to parse issuer CA certificate PEM")
+
+				return err
+			}
+
+			cert, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				skip.Log("tenant", t.Name, "err", fmt.Sprintf("failed to parse issuer certificate: %v", err))
+
+				return err
+			}
+
+			t.OIDC.issuerCA = cert
+		}
+	}
+
+	return nil
+}
+
+// Onboard a tenant.
+func newTenant(cfg *config, tCfg *tenantsConfig, r *chi.Mux, t *tenant) {
```

Comment:
Nice! Let's maybe rename to:

```suggestion
func registerTenant(cfg *config, tCfg *tenantsConfig, r *chi.Mux, t *tenant) {
```